dsl/src/commonMain/kotlin/it/unibo/collektive/aggregate/api/impl/AggregateContext.kt

```diff
@@ -0,0 +1,135 @@
+package it.unibo.collektive.aggregate.api.impl
+
+import arrow.core.Option
+import arrow.core.getOrElse
+import arrow.core.some
+import it.unibo.collektive.ID
+import it.unibo.collektive.aggregate.api.Aggregate
+import it.unibo.collektive.aggregate.api.YieldingContext
+import it.unibo.collektive.aggregate.api.YieldingContext.YieldingResult
+import it.unibo.collektive.aggregate.api.impl.stack.Stack
+import it.unibo.collektive.aggregate.api.operators.neighbouring
+import it.unibo.collektive.field.Field
+import it.unibo.collektive.networking.InboundMessage
+import it.unibo.collektive.networking.OutboundMessage
+import it.unibo.collektive.networking.SingleOutboundMessage
+import it.unibo.collektive.path.Path
+import it.unibo.collektive.state.State
+import it.unibo.collektive.state.impl.getTyped
+
+/**
+ * Context for managing aggregate computation.
+ * It represents the [localId] of the device, the [messages] received from the neighbours,
+ * and the [previousState] of the device.
+ */
+internal class AggregateContext(
+    override val localId: ID,
+    private val messages: Iterable<InboundMessage>,
+    private val previousState: State,
+) : Aggregate {
+
+    private val stack = Stack()
+    private var state: State = mapOf()
+    private var toBeSent = OutboundMessage(localId, emptyMap())
+
+    /**
+     * Messages to send to the other nodes.
+     */
+    fun messagesToSend(): OutboundMessage = toBeSent
+
+    /**
+     * Return the current state of the device as a new state.
+     */
+    fun newState(): State = state
+
+    private fun <T> newField(localValue: T, others: Map<ID, T>): Field<T> = Field(localId, localValue, others)
+
+    override fun <X> exchange(initial: X, body: (Field<X>) -> Field<X>): Field<X> {
+        val messages = messagesAt<X>(stack.currentPath())
+        val previous = stateAt(stack.currentPath(), initial)
+        val subject = newField(previous, messages)
+        return body(subject).also { field ->
+            val message = SingleOutboundMessage(field.localValue, field.excludeSelf())
+            val path = stack.currentPath()
+            check(!toBeSent.messages.containsKey(path)) {
+                """
+                    Alignment was broken by multiple aligned calls with the same path: $path.
+                    The most likely cause is an aggregate function call within a loop
+                """.trimIndent()
```

Comment:
can we add a test in which we just do something like:
repeat(2) { neighboring(0) } expecting to see this exception?

Also I wonder if we can intercept some of these in the compiler plugin, can we open an issue?

Reply:
I will open an issue and a separate PR introducing that test.